File: backend/src/main/java/cn/etarch/mao/harness/tool/FileChangeDiffUtil.java
package cn.etarch.mao.harness.tool;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;

import java.nio.charset.StandardCharsets;
import java.util.LinkedHashMap;
import java.util.Map;

public final class FileChangeDiffUtil {

    public static final String PRIVATE_DIFF_FIELD = "file_change_diff";
    public static final int SNAPSHOT_LIMIT_BYTES = 512 * 1024;
    public static final int PATCH_LIMIT_CHARS = 256 * 1024;

    private static final int PATCH_CONTEXT_LINES = 3;
    private static final int LCS_CELL_LIMIT = 2_000_000;

    private FileChangeDiffUtil() {
    }

    public static Map<String, Object> buildDiff(String path, String beforeContent, String afterContent) {
        String before = beforeContent != null ? beforeContent : "";
        String after = afterContent != null ? afterContent : "";
        Map<String, Object> diff = new LinkedHashMap<>();

        if (isBinary(before) || isBinary(after)) {
            diff.put("diff_mode", "UNSUPPORTED");
            diff.put("diff_unavailable_reason", "二进制文件无法生成文本 diff");
            diff.put("patch_truncated", false);
            return diff;
        }

        if (utf8Bytes(before) <= SNAPSHOT_LIMIT_BYTES && utf8Bytes(after) <= SNAPSHOT_LIMIT_BYTES) {
            diff.put("diff_mode", "SNAPSHOT");
            diff.put("before_content", before);
            diff.put("after_content", after);
            diff.put("patch_truncated", false);
            return diff;
        }

        PatchResult patch = buildUnifiedPatch(path, before, after);
        diff.put("diff_mode", "PATCH");
        diff.put("patch_content", patch.content());
        diff.put("patch_truncated", patch.truncated());
        return diff;
    }

    public static String stripPrivateDiff(String result, ObjectMapper objectMapper) {
        if (result == null || result.isBlank()) {
            return result;
        }
        try {
            JsonNode node = objectMapper.readTree(result);
            if (node instanceof ObjectNode objectNode && objectNode.has(PRIVATE_DIFF_FIELD)) {
                objectNode.remove(PRIVATE_DIFF_FIELD);
                return objectMapper.writeValueAsString(objectNode);
            }
        } catch (Exception ignored) {
        }
        return result;
    }

    public static LineDelta computeLineDelta(String beforeContent, String afterContent) {
        String[] oldLines = splitLines(beforeContent);
        String[] newLines = splitLines(afterContent);
        if (oldLines.length == 0) {
            return new LineDelta(newLines.length, 0);
        }
        if (newLines.length == 0) {
            return new LineDelta(0, oldLines.length);
        }

        int prefix = 0;
        while (prefix < oldLines.length && prefix < newLines.length && oldLines[prefix].equals(newLines[prefix])) {
            prefix++;
        }

        int oldSuffix = oldLines.length - 1;
        int newSuffix = newLines.length - 1;
        while (oldSuffix >= prefix && newSuffix >= prefix && oldLines[oldSuffix].equals(newLines[newSuffix])) {
            oldSuffix--;
            newSuffix--;
        }

        int oldChanged = oldSuffix - prefix + 1;
        int newChanged = newSuffix - prefix + 1;
        if (oldChanged <= 0) {
            return new LineDelta(Math.max(0, newChanged), 0);
        }
        if (newChanged <= 0) {
            return new LineDelta(0, Math.max(0, oldChanged));
        }

        long cells = (long) oldChanged * (long) newChanged;
        if (cells > LCS_CELL_LIMIT) {
            return new LineDelta(newChanged, oldChanged);
        }

        int lcs = lcsLength(oldLines, prefix, oldSuffix, newLines, prefix, newSuffix);
        return new LineDelta(newChanged - lcs, oldChanged - lcs);
    }

    private static boolean isBinary(String text) {
        int checkLen = Math.min(text.length(), 8192);
        for (int i = 0; i < checkLen; i++) {
            if (text.charAt(i) == 0) {
                return true;
            }
        }
        return false;
    }

    private static int utf8Bytes(String text) {
        return text.getBytes(StandardCharsets.UTF_8).length;
    }

    private static PatchResult buildUnifiedPatch(String path, String before, String after) {
        String[] oldLines = splitLines(before);
        String[] newLines = splitLines(after);

        int prefix = 0;
        while (prefix < oldLines.length && prefix < newLines.length && oldLines[prefix].equals(newLines[prefix])) {
            prefix++;
        }

        int oldSuffix = oldLines.length - 1;
        int newSuffix = newLines.length - 1;
        while (oldSuffix >= prefix && newSuffix >= prefix && oldLines[oldSuffix].equals(newLines[newSuffix])) {
            oldSuffix--;
            newSuffix--;
        }

        int contextStart = Math.max(0, prefix - PATCH_CONTEXT_LINES);
        int oldContextEnd = Math.min(oldLines.length - 1, oldSuffix + PATCH_CONTEXT_LINES);
        int newContextEnd = Math.min(newLines.length - 1, newSuffix + PATCH_CONTEXT_LINES);

        int oldStartLine = contextStart + 1;
        int newStartLine = contextStart + 1;
        int oldCount = Math.max(0, oldContextEnd - contextStart + 1);
        int newCount = Math.max(0, newContextEnd - contextStart + 1);

        StringBuilder patch = new StringBuilder(Math.min(PATCH_LIMIT_CHARS, 8192));
        boolean truncated = false;
        truncated |= appendBounded(patch, "--- a/" + path + "\n");
        truncated |= appendBounded(patch, "+++ b/" + path + "\n");
        truncated |= appendBounded(patch, "@@ -" + oldStartLine + "," + oldCount
                + " +" + newStartLine + "," + newCount + " @@\n");

        for (int i = contextStart; i < prefix && i < oldLines.length; i++) {
            truncated |= appendBounded(patch, " " + oldLines[i] + "\n");
        }
        for (int i = prefix; i <= oldSuffix && i < oldLines.length; i++) {
            truncated |= appendBounded(patch, "-" + oldLines[i] + "\n");
        }
        for (int i = prefix; i <= newSuffix && i < newLines.length; i++) {
            truncated |= appendBounded(patch, "+" + newLines[i] + "\n");
        }

        int sharedTailStart = Math.max(prefix, Math.max(oldSuffix + 1, newSuffix + 1));
        int sharedTailEnd = Math.min(oldContextEnd, oldLines.length - 1);
        for (int i = sharedTailStart; i <= sharedTailEnd; i++) {
            truncated |= appendBounded(patch, " " + oldLines[i] + "\n");
        }

        if (truncated && !patch.toString().endsWith("\n...[diff truncated]\n")) {
            int maxPrefix = Math.max(0, PATCH_LIMIT_CHARS - "\n...[diff truncated]\n".length());
            patch.setLength(Math.min(patch.length(), maxPrefix));
            patch.append("\n...[diff truncated]\n");
        }
        return new PatchResult(patch.toString(), truncated);
    }

    private static String[] splitLines(String text) {
        if (text == null || text.isEmpty()) {
            return new String[0];
        }
        return text.split("\\R", -1);
    }

    private static boolean appendBounded(StringBuilder builder, String text) {
        if (builder.length() >= PATCH_LIMIT_CHARS) {
            return true;
        }
        int remaining = PATCH_LIMIT_CHARS - builder.length();
        if (text.length() <= remaining) {
            builder.append(text);
            return false;
        }
        builder.append(text, 0, remaining);
        return true;
    }

    private static int lcsLength(String[] oldLines, int oldStart, int oldEnd,
                                 String[] newLines, int newStart, int newEnd) {
        int newLen = newEnd - newStart + 1;
        int[] previous = new int[newLen + 1];
        int[] current = new int[newLen + 1];

        for (int i = oldStart; i <= oldEnd; i++) {
            for (int j = 1; j <= newLen; j++) {
                if (oldLines[i].equals(newLines[newStart + j - 1])) {
                    current[j] = previous[j - 1] + 1;
                } else {
                    current[j] = Math.max(previous[j], current[j - 1]);
                }
            }
            int[] temp = previous;
            previous = current;
            current = temp;
            java.util.Arrays.fill(current, 0);
        }
        return previous[newLen];
    }

    private record PatchResult(String content, boolean truncated) {
    }

    public record LineDelta(int linesAdded, int linesDeleted) {
    }
}

File: backend-ts/src/harness/tool/file-change-diff-util.ts
export const PRIVATE_DIFF_FIELD = 'file_change_diff';
export const SNAPSHOT_LIMIT_BYTES = 512 * 1024;
export const PATCH_LIMIT_CHARS = 256 * 1024;
const PATCH_CONTEXT_LINES = 3;
const LCS_CELL_LIMIT = 2_000_000;

export interface LineDelta {
  linesAdded: number;
  linesDeleted: number;
}

export const FileChangeDiffUtil = {
  PRIVATE_DIFF_FIELD,
  SNAPSHOT_LIMIT_BYTES,
  PATCH_LIMIT_CHARS,
  buildDiff(filePath: string, beforeContent: string | null | undefined, afterContent: string | null | undefined): Record<string, unknown> {
    const before = beforeContent ?? '';
    const after = afterContent ?? '';
    const diff: Record<string, unknown> = {};

    if (isBinary(before) || isBinary(after)) {
      diff.diff_mode = 'UNSUPPORTED';
      diff.diff_unavailable_reason = '二进制文件无法生成文本 diff';
      diff.patch_truncated = false;
      return diff;
    }

    if (utf8Bytes(before) <= SNAPSHOT_LIMIT_BYTES && utf8Bytes(after) <= SNAPSHOT_LIMIT_BYTES) {
      diff.diff_mode = 'SNAPSHOT';
      diff.before_content = before;
      diff.after_content = after;
      diff.patch_truncated = false;
      return diff;
    }

    const patch = buildUnifiedPatch(filePath, before, after);
    diff.diff_mode = 'PATCH';
    diff.patch_content = patch.content;
    diff.patch_truncated = patch.truncated;
    return diff;
  },

  stripPrivateDiff(result: string | null | undefined): string | null | undefined {
    if (result == null || result.trim() === '') return result;
    try {
      const node = JSON.parse(result) as Record<string, unknown>;
      if (node && typeof node === 'object' && !Array.isArray(node) && PRIVATE_DIFF_FIELD in node) {
        delete node[PRIVATE_DIFF_FIELD];
        return JSON.stringify(node);
      }
    } catch {
      // ignore
    }
    return result;
  },

  computeLineDelta(beforeContent: string | null | undefined, afterContent: string | null | undefined): LineDelta {
    const oldLines = splitLines(beforeContent);
    const newLines = splitLines(afterContent);
    if (oldLines.length === 0) return { linesAdded: newLines.length, linesDeleted: 0 };
    if (newLines.length === 0) return { linesAdded: 0, linesDeleted: oldLines.length };

    let prefix = 0;
    while (prefix < oldLines.length && prefix < newLines.length && oldLines[prefix] === newLines[prefix]) prefix++;

    let oldSuffix = oldLines.length - 1;
    let newSuffix = newLines.length - 1;
    while (oldSuffix >= prefix && newSuffix >= prefix && oldLines[oldSuffix] === newLines[newSuffix]) {
      oldSuffix--;
      newSuffix--;
    }

    const oldChanged = oldSuffix - prefix + 1;
    const newChanged = newSuffix - prefix + 1;
    if (oldChanged <= 0) return { linesAdded: Math.max(0, newChanged), linesDeleted: 0 };
    if (newChanged <= 0) return { linesAdded: 0, linesDeleted: Math.max(0, oldChanged) };

    const cells = oldChanged * newChanged;
    if (cells > LCS_CELL_LIMIT) return { linesAdded: newChanged, linesDeleted: oldChanged };

    const lcs = lcsLength(oldLines, prefix, oldSuffix, newLines, prefix, newSuffix);
    return { linesAdded: newChanged - lcs, linesDeleted: oldChanged - lcs };
  },
};

function isBinary(text: string): boolean {
  const checkLen = Math.min(text.length, 8192);
  for (let i = 0; i < checkLen; i++) {
    if (text.charCodeAt(i) === 0) return true;
  }
  return false;
}

function utf8Bytes(text: string): number {
  return Buffer.byteLength(text, 'utf8');
}

function buildUnifiedPatch(filePath: string, before: string, after: string): { content: string; truncated: boolean } {
  const oldLines = splitLines(before);
  const newLines = splitLines(after);

  let prefix = 0;
  while (prefix < oldLines.length && prefix < newLines.length && oldLines[prefix] === newLines[prefix]) prefix++;

  let oldSuffix = oldLines.length - 1;
  let newSuffix = newLines.length - 1;
  while (oldSuffix >= prefix && newSuffix >= prefix && oldLines[oldSuffix] === newLines[newSuffix]) {
    oldSuffix--;
    newSuffix--;
  }

  const contextStart = Math.max(0, prefix - PATCH_CONTEXT_LINES);
  const oldContextEnd = Math.min(oldLines.length - 1, oldSuffix + PATCH_CONTEXT_LINES);
  const newContextEnd = Math.min(newLines.length - 1, newSuffix + PATCH_CONTEXT_LINES);

  const oldStartLine = contextStart + 1;
  const newStartLine = contextStart + 1;
  const oldCount = Math.max(0, oldContextEnd - contextStart + 1);
  const newCount = Math.max(0, newContextEnd - contextStart + 1);

  let patch = '';
  let truncated = false;
  const append = (text: string) => {
    if (patch.length >= PATCH_LIMIT_CHARS) {
      truncated = true;
      return;
    }
    const remaining = PATCH_LIMIT_CHARS - patch.length;
    if (text.length <= remaining) {
      patch += text;
    } else {
      patch += text.slice(0, remaining);
      truncated = true;
    }
  };

  append(`--- a/${filePath}\n`);
  append(`+++ b/${filePath}\n`);
  append(`@@ -${oldStartLine},${oldCount} +${newStartLine},${newCount} @@\n`);

  for (let i = contextStart; i < prefix && i < oldLines.length; i++) append(` ${oldLines[i]}\n`);
  for (let i = prefix; i <= oldSuffix && i < oldLines.length; i++) append(`-${oldLines[i]}\n`);
  for (let i = prefix; i <= newSuffix && i < newLines.length; i++) append(`+${newLines[i]}\n`);

  // 尾部上下文：统一按旧序号从 oldSuffix+1 起输出。共享尾段在新旧文件中内容相同、仅索引错位
  // （纯插入时 newSuffix 大于 oldSuffix），若取 max(oldSuffix+1, newSuffix+1) 会按旧序号跳过共享行。
  const sharedTailStart = Math.max(prefix, oldSuffix + 1);
  const sharedTailEnd = Math.min(oldContextEnd, oldLines.length - 1);
  for (let i = sharedTailStart; i <= sharedTailEnd; i++) append(` ${oldLines[i]}\n`);

  if (truncated && !patch.endsWith('\n...[diff truncated]\n')) {
    const marker = '\n...[diff truncated]\n';
    const maxPrefix = Math.max(0, PATCH_LIMIT_CHARS - marker.length);
    patch = patch.slice(0, Math.min(patch.length, maxPrefix)) + marker;
  }
  return { content: patch, truncated };
}

function splitLines(text: string | null | undefined): string[] {
  if (text == null || text === '') return [];
  return text.split(/\r\n|\n|\r/);
}

function lcsLength(oldLines: string[], oldStart: number, oldEnd: number, newLines: string[], newStart: number, newEnd: number): number {
  const newLen = newEnd - newStart + 1;
  let previous = new Array<number>(newLen + 1).fill(0);
  let current = new Array<number>(newLen + 1).fill(0);

  for (let i = oldStart; i <= oldEnd; i++) {
    for (let j = 1; j <= newLen; j++) {
      if (oldLines[i] === newLines[newStart + j - 1]) {
        current[j] = previous[j - 1] + 1;
      } else {
        current[j] = Math.max(previous[j], current[j - 1]);
      }
    }
    const temp = previous;
    previous = current;
    current = temp;
    current.fill(0);
  }
  return previous[newLen];
}
